fedora/client/openidcclient.py

```diff
@@ -331,32 +311,68 @@ def token_app(environ, start_response):
-        the token has expired.
+        with the current app_identifier that has the required scopes, we weill
+        return it.
 
-        This function will return a valid bearer token or throw an exception.
+        This function will return a bearer token or None.
+        Note that the bearer token might have been revoked by the user or
+        expired.
+        In that case, you will want to call report_token_issue() to try to
+        renew the token or delete the token.
 
         :kwarg scopes: A list of scopes required for the current client.
         :kwarg new_token: If True, we will actively request the user to get a
             new token with the current scopeset if we do not already have on.
         """
-        token = self._get_valid_token_with_scopes(scopes)
+        if not isinstance(scopes, list):
+            raise Exception('Scopes must be a list')
+        token = self._get_token_with_scopes(scopes)
         if token:
             # If we had a valid token, use that
-            return token['access_token']
+            self.last_returned_uuid = token[0]
+            self.problem_reported = False
+            return token[1]['access_token']
         elif not new_token:
             return None
 
         # We did not have a valid token, now comes the hard part...
         uuid = self._get_new_token(scopes)
         if uuid:
+            self.last_returned_uuid = uuid
+            self.problem_reported = False
             return self._cache[uuid]['access_token']
 
-    def send_request(self, method, scopes=None, verb='POST', new_token=True,
+    def report_token_issue(self):
+        """Report an error with the last token that was returned.
+
+        This will attempt to renew the token that was last returned.
+        If that worked, we will return the new access token.
+        If it did not work, we will return None and remove this token from the
+        cache.
+
+        If you get an indication from your application that the token you sent
+        was invalid, you should call it.
+        You should explicitly NOT call this function if the token was valid but
+        your request failed due to a server error or because the account or
+        token was lacking specific permissions.
+        """
+        if not self.last_returned_uuid:
+            raise Exception('Cannot report issue before requesting token')
+        if self.problem_reported:
+            # We were reported an issue before. Let's just remove this token.
+            self._delete_token(self.last_returned_uuid)
+            return None
+        refresh_result = self._refresh_token(self.last_returned_uuid)
+        if not refresh_result:
+            self._delete_token(self.last_returned_uuid)
+            return None
+        else:
+            self.problem_reported = True
+            return self._cache[self.last_returned_uuid]['access_token']
+
+    def send_request(self, method, scopes, verb='POST', new_token=True,
```

Comment:
A few comments here, even though you haven't changed this.

* ``method`` does not feel like an appropriate name to me. When I think method and HTTP, I think [RFC2616](https://tools.ietf.org/html/rfc2616#section-5.1.1) method, e.g. OPTIONS, GET, HEAD, POST, etc. ``path`` might be better. It's even worse because you use the ``verb`` keyword argument and then use it as the ``requests.request`` method argument.

* From your documentation, it's not clear you're using ``requests`` so I don't know what should be in the ``kwargs`` dictionary until I realize that by reading the code.

* The above isn't right because you're not actually giving those kwargs to requests, you're providing them as the ``data`` keyword argument in requests, which form-encodes that dictionary.

I _highly_ recommend simply wrapping the ``requests.request`` API here and popping whatever keyword arguments you want this function to have off ``**kwargs`` before throwing them into ``requests.request``. It's a much cleaner API.